opus/application/static_media/js/browse.js

```diff
@@ -548,34 +521,94 @@ var o_browse = {
                 // the || is for cross-browser support; firefox does not support keyCode
                 switch (e.which || e.keyCode) {
                     case 32:  // spacebar
-                        if (opus.metadataDetailOpusId !== "") {
+                        if (opus.metadataDetailOpusId !== "" && opus.metadataDetailOpusId !== undefined && !$("#galleryViewContents").hasClass("op-disabled")) {
                             o_browse.undoRangeSelect();
                             o_cart.toggleInCart(opus.metadataDetailOpusId);
                         }
                         break;
                     case 39:  // next
-                        detailOpusId = $("#galleryView").find(".op-next").data("id");
-                        o_browse.removeEditMetadataDetails();
-                        o_browse.loadPageIfNeeded("next", detailOpusId);
+                        obsNum++;
+                        // if this new observation is in the last row of cached observations, don't allow the user to arrow there
+                        if (lastCachedObsNum === viewNamespace.totalObsCount || obsNum + viewNamespace.galleryBoundingRect.x <= lastCachedObsNum) {
+                            detailOpusId = $("#galleryView").find(".op-next").data("id");
+                            o_browse.removeEditMetadataDetails();
+                            o_browse.loadPageIfNeeded("next", detailOpusId);
+                        } else {
+                            return false;
+                        }
                         break;
                     case 37:  // prev
+                        obsNum--;
                         detailOpusId = $("#galleryView").find(".op-prev").data("id");
                         o_browse.removeEditMetadataDetails();
                         o_browse.loadPageIfNeeded("prev", detailOpusId);
                         break;
+                    case 38:  // up
+                        offset = (o_browse.isGalleryView() ? viewNamespace.galleryBoundingRect.x : 1);
+                        if (obsNum > offset) {
+                            obsNum -= offset;
+                            detailOpusId = (o_browse.isGalleryView() ? $(tab).find(`.op-thumbnail-container[data-obs='${obsNum}']`).data("id") : $(tab).find(`tr[data-obs='${obsNum}']`).data("id"));
+                            o_browse.removeEditMetadataDetails();
+                            o_browse.loadPageIfNeeded("prev", detailOpusId);
+                        }
+                        break;
+                    case 40:  // down
+                        offset = (o_browse.isGalleryView() ? viewNamespace.galleryBoundingRect.x : 1);
+                        if (obsNum + offset <= viewNamespace.totalObsCount) {
+                            obsNum += offset;
+                            // if this new observation is in the last row of cached observations, don't allow the user to arrow there
```

Comment:
Same comment as above about this comment and only doing on load in progress.

Reply:
comment clarified in both places.